Add color to custom logger

borkaehw opened the pull request:
[Warn], [Error], and [Info] will be shown in yellow, red, and green respectively.

andyscott commented:
Is there any way to make this respect the `--color` flag? https://docs.bazel.build/versions/master/command-line-reference.html#flag--color

andyscott commented on src/main/scala/higherkindness/rules_scala/workers/zinc/common/ColorMessage.scala:
```diff
@@ -0,0 +1,28 @@
+import xsbti.Problem
+import Console.{RED => CR, YELLOW => CY, GREEN => CG, RESET => CRESET}
+
+class ColorMessage() {
```
Let's just make this a singleton `object`.

borkaehw replied:
Good idea, fixed.

andyscott commented on rules/scalafmt/scalafmt/ScalafmtRunner.scala:
```diff
@@ -35,12 +36,14 @@ object ScalafmtRunner extends WorkerMain[Unit] {
       if (code == formatted) code else format(formatted)
     }
 
+    val colorMessage = new ColorMessage()
```
This isn't needed if we make `ColorMessage` a singleton.

borkaehw replied:
Good idea, fixed.

andyscott commented on src/main/scala/higherkindness/rules_scala/workers/zinc/common/ColorMessage.scala:
```diff
@@ -0,0 +1,28 @@
+
+class ColorMessage() {
+  def CProblem(problem: Problem, color: String): String = {
+    val CC = color match {
```
Pattern matching on a string seems like a bad idea. How about we break this out into separate methods according to color?

Usage then looks something like:
```
Color.formatYellow("my message")
Color.formatRed("a red message")
```

borkaehw replied:
I personally think `formatWarning` makes more sense.

andyscott commented on src/main/scala/higherkindness/rules_scala/workers/zinc/common/ColorMessage.scala:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  def CString(message: String, color: String): String = {
+    val CCSeverity = color match {
```
Same thing as before. Additionally, the message prefix should determine color (here you have color determining prefix).

This can be broken out into methods for error, warn, and info.

borkaehw replied:
Good idea, fixed.

andyscott commented on src/main/scala/higherkindness/rules_scala/workers/zinc/common/LoggedReporter.scala:
```diff
@@ -0,0 +1,135 @@
+// The following code is based on scala.tools.nsc.reporters.{AbstractReporter, ConsoleReporter, Reporter}
+// Copyright 2002-2009 LAMP/EPFL
+// see licenses/LICENSE_Scala
+// Original author: Martin Odersky
```
How much of this code is from scala.tools.nsc.reporters? Perhaps we can move this to the thirdparty directory and add an appropriate notice.

borkaehw replied:
https://github.com/sbt/zinc/blob/develop/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/LoggedReporter.scala
This is where we took the file from. We made some changes so they aren't exactly the same.

borkaehw replied:
Perhaps we could remove this comment? While, 98 101 105 106 107 are the only lines that I changed.

andyscott replied:
As it stands, merely copying in this code is a clear violation of the original license(s). We'll need to follow the guidelines for the original source of this code from the Scala compiler _and_ and any additional constraints/licensing for the derivative work in Zinc. 

We have two options, as far as I see it:

- If you want to add this code to this repo we will need to provide a copy of Scalac and/or Zinc's LICENSE, with this derived code clearly attributed to those sources and with references to the licenses. We must follow the rules set out for the code we're copying in.

- Alternatively we can write our own reporter logic and avoid any OSS license SNAFUs.

jjudd replied:
I'm all for writing our own. It's not a ton of code and doesn't seem too complex. What are people's thoughts about that? 

I'd rather avoid any license SNAFUs if we can.

andyscott replied:
I'd prefer that we write our own, but I won't block this PR on preference alone.

borkaehw replied:
I also prefer to write our own. We could have more flexibility later on. I will work on it.

borkaehw commented:
I will investigate more on `--color` flag.

borkaehw commented:
@andyscott It seems like we don't need to treat `--color` separately. `--color` disables colors both on Bazel output and worker output. I am not sure how it works under the hood but that is what I see.

I also fixed the CI and it's ready for review again.

borkaehw commented:
@andyscott friendly ping here.

borkaehw commented:
@andyscott I have rewritten the logger to make it less like copying, but the file is too simple to make a huge difference. I removed some parts that we didn't utilize, and added tests for our own logger. Please take a look, thanks.

andyscott commented:
This looks better, but the file is still a clear derivative of another project's code, so it's subject to the same licensing problems (as far as I know).

You'll need to write the colored logging code from scratch _without basing your code off of existing sources or referencing those files to guide your development_.

borkaehw commented:
@andyscott I rewrote the logger again. Let me know if you think it's still not distinguish enough.

andyscott commented:
This looks great. Thanks!
